ddsrouter_core/src/cpp/efficiency/FastPayloadPool.cpp

```diff
@@ -0,0 +1,156 @@
+}
+
+bool FastPayloadPool::get_payload(
+        const Payload& src_payload,
+        IPayloadPool*& data_owner,
+        Payload& target_payload)
+{
+    // If we are not the owner, create a new payload. Else, reference the existing one
+    if (data_owner != this)
+    {
+        // Store space for payload
+        if (!get_payload(src_payload.max_size, target_payload))
+        {
+            return false;
+        }
+
+        // Copy info
+        std::memcpy(target_payload.data, src_payload.data, src_payload.length);
+        target_payload.length = src_payload.length;
+    }
+    else
+    {
+        // IMPORTANT: If payload has been reserved from this object, it must have the +4 bytes before data
+        // We get this value and add a +1 to set that is referencing from one more payload
+        MetaInfoType* reference_place = reinterpret_cast<MetaInfoType*>(src_payload.data);
+        reference_place--;
+
+        // Add reference
+        (*reference_place)++;
+
+        // Set Payload to refer same payload
+        target_payload.data = src_payload.data;
+        target_payload.length = src_payload.length;
+        target_payload.max_size = src_payload.max_size;
+    }
+    return true;
+}
+
+bool FastPayloadPool::release_payload(
+        Payload& payload)
+{
+    // IMPORTANT: If payload has been reserved from this object, it must have the +4 bytes before data
+    // We get this value and add a -1 to set that is referencing from one less payload
+    MetaInfoType* reference_place = reinterpret_cast<MetaInfoType*>(payload.data);
+    reference_place--;
+
+    // Remove reference
+    (*reference_place)--;
+
+    // In case it was the last reference, release payload
+    if ((*reference_place) == 0)
+    {
+        if (!release_(payload))
+        {
+            return false;
+        }
+    }
+
+    payload.length = 0;
+    payload.max_size = 0;
+    payload.data = nullptr;
+    payload.pos = 0;
+
+    return true;
+}
+
+bool FastPayloadPool::reserve_(
+        uint32_t size,
+        types::Payload& payload)
+{
+    if (size == 0)
+    {
+        logDevError(DDSROUTER_PAYLOADPOOL,
+                "Trying to reserve a data block of 0 bytes.");
+        return false;
+    }
+
+    // Allocate memory + 4 bytes for reference
+    void* memory_allocated = std::malloc(size + sizeof(MetaInfoType));
+
+    // Use reference space to set that this is referenced for the first time
+    MetaInfoType* reference_place = reinterpret_cast<MetaInfoType*>(memory_allocated);
+    (*reference_place) = 1;
+
+    payload.data = reinterpret_cast<eprosima::fastrtps::rtps::octet *>(reference_place + 1);
+    payload.max_size = size;
+
+    add_reserved_payload_();
+
+    return true;
+}
+
+bool FastPayloadPool::release_(
+        types::Payload& payload)
+{
+    // Free memory from the initial allocation, 4 bytes before
+    MetaInfoType* reference_place = reinterpret_cast<MetaInfoType*>(payload.data);
+    reference_place--;
+    free(reference_place);
```

Comment:
What do you think about encapsulating the malloc and free functions so that we can handle the exceptions that could raise from these.

Reply:
encapsulating in PayloadPool or along the whole code? We do not use them but here, but could be a good idea :ok: 
That's the first step for implementing a garbage collector :smiling_imp:

Reply:
Let's leave this for future improvements then